RHIDP-12611: Document the RHDH must-gather tool for diagnostics data collection (#2151)
redhat-developer/red-hat-developers-documentation-rhdh: assemblies/observe_diagnostic-data-collection/assembly-generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases.adoc

:_mod-docs-content-type: ASSEMBLY
ifdef::context[:parent-context: {context}]

[id="generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases_{context}"]
= Generate diagnostic data using the must-gather tool to streamline support cases

:context: generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases

[role="_abstract"]
When opening a support ticket for {product-very-short} deployment issues, use the must-gather tool to quickly generate diagnostic data.
The tool collects deployment manifests, logs, and resource definitions in one operation to resolve platform issues faster.

Run must-gather on {ocp-short} using `oc adm must-gather`. On Kubernetes platforms ({aks-short}, {eks-short}, {gke-short}), deploy a Helm chart.

== When to use must-gather

Use must-gather when:

* Opening a support ticket with {company-name} Global Support Services
* Troubleshooting deployment, configuration, or runtime issues

> * Conducting health checks before upgrades

Reviewer: Maybe `Capturing deployment state before RHDH upgrades` would be more accurate, as the tool itself doesn't analyze or evaluate health? It just collects data without analysis.

:_mod-docs-content-type: ASSEMBLY
ifdef::context[:parent-context: {context}]

[id="generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases_{context}"]
= Generate diagnostic data using the must-gather tool to streamline support cases

:context: generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases

[role="_abstract"]

> Generate diagnostic data by using the must-gather tool to streamline support cases by collecting comprehensive diagnostic data in one operation to reduce resolution time and provide complete information to {company-name} support teams.

Reviewer 2: The sentence is redundant — it says "diagnostic data" twice and restates the same benefit (streamlining/reducing resolution time/providing complete information) three different ways. Here is a simplified option:

```suggestion
Collect comprehensive diagnostic data in one operation by using the must-gather tool to provide complete information to the support team and reduce investigation time.
```

:_mod-docs-content-type: ASSEMBLY
ifdef::context[:parent-context: {context}]

[id="generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases_{context}"]
= Generate diagnostic data using the must-gather tool to streamline support cases

:context: generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases

[role="_abstract"]
When opening a support case for {product-very-short} deployment issues, providing diagnostic data accelerates troubleshooting and reduces resolution time.
The must-gather tool collects deployment manifests, logs, and resource definitions in one operation.

Run must-gather on {ocp-short} using `oc adm must-gather`, or on Kubernetes platforms ({aks-short}, {eks-short}, {gke-short}) by deploying a Helm chart.
The tool collects data from any {product-very-short} instance regardless of deployment version.
Use the `latest` image tag to ensure access to the most recent collection capabilities and fixes.

Use the {product-very-short} must-gather tool to:

* Prepare support cases with {company-name} Global Support Services
* Troubleshoot production incidents
* Conduct proactive health checks and configuration validation

> * Debug deployment issues in development or staging environments

Reviewer: I feel like bullets 2 and 4 say the same thing, i.e., troubleshooting problems, just in different environments. IMO, listing "production" and "development or staging" separately implies a meaningful distinction in how the tool works, but there isn't one. The must-gather image collects the same RHDH-related data regardless of the environment.
Maybe collapse these 2 bullets into something like `Troubleshoot deployment, configuration, or runtime issues`?

Run must-gather on {ocp-short} using `oc adm must-gather`, or on Kubernetes platforms ({aks-short}, {eks-short}, {gke-short}) by deploying a Helm chart.

Use the {product-very-short} must-gather tool to:

* Prepare support cases with {company-name} Global Support Services
* Troubleshoot production incidents
* Conduct proactive health checks and configuration validation
* Debug deployment issues in development or staging environments

include::../../artifacts/snip-technology-preview.adoc[]

include::../modules/observe_diagnostic-data-collection/ref-diagnostic-data-types-and-collection-scope.adoc[leveloffset=+1]

include::../modules/observe_diagnostic-data-collection/proc-run-the-rhdh-must-gather-image-on-openshift-to-collect-deployment-manifests-and-logs.adoc[leveloffset=+1]

include::../modules/observe_diagnostic-data-collection/proc-run-the-rhdh-must-gather-image-on-kubernetes-to-collect-deployment-manifests-and-logs.adoc[leveloffset=+1]

include::../modules/observe_diagnostic-data-collection/proc-filter-data-scope-on-openshift-to-reduce-collection-time.adoc[leveloffset=+1]

> include::../modules/observe_diagnostic-data-collection/proc-filter-data-scope-on-kubernetes-to-reduce-collection-time.adoc[leveloffset=+1]

Reviewer 2: These 2 procedures are repeating in another form the content already in observe_diagnostic-data-collection/proc-run-the-rhdh-must-gather-image-on-openshift-to-collect-deployment-manifests-and-logs.adoc proc-run-the-rhdh-must-gather-image-on-kubernetes-to-collect-deployment-manifests-and-logs.adoc. 

Consider consolidating in 1 procedure for Operator-based deployments and 1 procedure for Helm-based deployments.

Additionally, it is possible to install the Operator on Kubernetes as well as OpenShift, therefore it would be more accurate to focus the procedure titles on the deployment method.

include::../modules/observe_diagnostic-data-collection/proc-collect-heap-dumps-with-the-inspector-protocol-to-diagnose-memory-issues.adoc[leveloffset=+1]

include::../modules/observe_diagnostic-data-collection/proc-collect-heap-dumps-with-custom-sigusr2-configuration-to-diagnose-memory-issues.adoc[leveloffset=+1]

> include::../modules/observe_diagnostic-data-collection/ref-configuration-options-for-diagnostic-collection.adoc[leveloffset=+1]

Reviewer 2: Avoid repeating  what the user has read before. Consider consolidating the information in one single location.

Command-line flags should go in the procedure for Operator-based deployments.

Environment variables and Helm chart values => should go in the procedure for Helm-based deployments.

Platform-specific configuration => should go in the introduction for each platform-specific procedure.

 Version compatibility => should go in the general introduction.

:context: generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases

[role="_abstract"]
When opening a support ticket for {product-very-short} deployment issues, use the must-gather tool to quickly generate diagnostic data.
The tool collects deployment manifests, logs, and resource definitions in one operation to resolve platform issues faster.

Run must-gather on {ocp-short} using `oc adm must-gather`. On Kubernetes platforms ({aks-short}, {eks-short}, {gke-short}), deploy a Helm chart.

== When to use must-gather

Use must-gather when:

* Opening a support ticket with {company-name} Global Support Services
* Troubleshooting production incidents
* Conducting health checks before upgrades
* Debugging deployment issues

[NOTE]
====
Typical collection size:

* Basic collection: 10-50 MB
* With heap dumps: 500 MB - 2 GB

Plan for adequate storage and bandwidth when collecting for support tickets.
====

include::../modules/observe_diagnostic-data-collection/con-deployment-methods-and-data-collection.adoc[leveloffset=+1]

include::../modules/observe_diagnostic-data-collection/ref-diagnostic-data-types-and-collection-scope.adoc[leveloffset=+1]

> include::../modules/observe_diagnostic-data-collection/proc-run-must-gather-for-operator-based-deployments-to-collect-diagnostic-data.adoc[leveloffset=+1]

Reviewer: Before elaborating on all these (advanced) use cases below, I feel like we are missing a simple Quick Start section to "just" deploy the must-gather with its default options and how to fetch the data.
Like I just have access to a cluster and I just want to collect data about all the running RHDH instances (regardless of the deployment method).
See https://github.com/redhat-developer/rhdh-must-gather/tree/release-1.10#quick-start

:_mod-docs-content-type: ASSEMBLY
ifdef::context[:parent-context: {context}]

[id="generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases_{context}"]
= Generate diagnostic data using the must-gather tool to streamline support cases

:context: generate-diagnostic-data-using-the-must-gather-tool-to-streamline-support-cases

[role="_abstract"]
When opening a support ticket for {product-very-short} deployment issues, use the must-gather tool to quickly generate diagnostic data.
The tool collects deployment manifests, logs, and resource definitions in one operation to resolve platform issues faster.

Use must-gather when opening a support ticket with {company-name} Global Support Services, troubleshooting deployment issues, or capturing deployment state before {product-very-short} upgrades.

> *On {ocp-short}:*

Reviewer 2: ```suggestion
```

[source,terminal]
----
$ oc adm must-gather --image=registry.access.redhat.com/rhdh/rhdh-must-gather-rhel9:latest
----

Output: `./must-gather.local.<timestamp>`

*On supported Kubernetes platforms:*

[source,terminal]
----
$ helm install my-rhdh-must-gather redhat-developer-hub-must-gather \
  --repo https://charts.openshift.io \
  --namespace rhdh-diagnostics \
  --create-namespace

$ kubectl wait --for=condition=ready pod \
  -l app.kubernetes.io/instance=my-rhdh-must-gather,app.kubernetes.io/component=gather \
  --timeout=3600s -n rhdh-diagnostics

$ kubectl exec deploy/my-rhdh-must-gather -c data-holder -n rhdh-diagnostics -- \
  tar czf - -C /must-gather . > rhdh-must-gather-output.tar.gz

$ helm uninstall my-rhdh-must-gather -n rhdh-diagnostics

> ----

Reviewer 2: Remove. This procedure should explain exclusively how to run `oc adm must-gather`. Running with helm is *the other procedure*.

```suggestion
```
